File: Templates/TypeEditor/src/Templates/Templates/VB.TypeEditor/CustomControlLibrary.Server/CustomTypeEditor/Handler/CreateCustomTypeEditorVMHandler.vb
Imports CustomControlLibrary.Protocol
Imports CustomControlLibrary.Protocol.Endpoints
Imports Microsoft.DotNet.DesignTools.Protocol.Endpoints

Namespace Handlers
    ''' <summary>
    ''' The handler for <see cref="CreateCustomTypeEditorVMEndpoint"/>. 
    ''' This actually creates the ViewModel and returns it via its request class.
    ''' </summary>
    <ExportRequestHandler(EndpointNames.CreateCustomTypeEditorVM)>
    Friend Class CreateCustomTypeEditorVMHandler
        Inherits RequestHandler(Of CreateCustomTypeEditorVMRequest, CreateCustomTypeEditorVMResponse)

        Public Overrides Function HandleRequest(ByVal request As CreateCustomTypeEditorVMRequest) As CreateCustomTypeEditorVMResponse
            ' Gets the respective DesignerHost of the sessionId, which has been passed by the client.
            Dim designerHost = GetDesignerHost(request.SessionId)

            ' Creates the ViewModel and passes the DesignerHost.
            Dim viewModel = CreateViewModel(Of CustomTypeEditorVM)(designerHost)

            ' The ViewModel then initializes and wraps itself into the response class
            ' so it can be returned to the client.
            Return viewModel.Initialize(request.CustomControlProxy)
        End Function
    End Class
End Namespace

File: Templates/TypeEditor/src/TemplateSolutions/CS.CustomTypeEditor/CustomControlLibrary.Server/CustomTypeEditor/Handler/CreateCustomTypeEditorVMHandler.cs
using CustomControlLibrary.Protocol;
using CustomControlLibrary.Protocol.Endpoints;
using Microsoft.DotNet.DesignTools.Protocol.Endpoints;

namespace CustomControlLibrary.Designer.Server.Handlers
{
    /// <summary>
    /// The handler for <see cref="CreateCustomTypeEditorVMEndpoint"/>. 
    /// This actually creates the ViewModel and returns it via its request class.
    /// </summary>
    [ExportRequestHandler(EndpointNames.CreateCustomTypeEditorVM)]
    internal class CreateCustomTypeEditorVMHandler 
        : RequestHandler<CreateCustomTypeEditorVMRequest, CreateCustomTypeEditorVMResponse>
    {
        public override CreateCustomTypeEditorVMResponse HandleRequest(CreateCustomTypeEditorVMRequest request)
        {
            // Gets the respective DesignerHost of the sessionId, which has been passed by the client.
            var designerHost = GetDesignerHost(request.SessionId);

            // Creates the ViewModel and passes the DesignerHost.
            var viewModel = CreateViewModel<CustomTypeEditorVM>(designerHost);
            
            // The ViewModel then initializes and wraps itself into the response class
            // so it can be returned to the client.
            return viewModel.Initialize(request.CustomControlProxy!);
        }
    }
}
